modules/mod-deep-learning/ModelManagerPanel.cpp

```diff
@@ -0,0 +1,335 @@
+
+   // if this is the first card we're adding, go ahead and select it
+   if (mPanels.size() == 1)
+   {
+      mEffect->SetModel(card);
+   }
+
+   ShuttleGui S(mScroller, eIsCreating);
+   S.AddWindow(mPanels[repoId].get(), wxEXPAND);
+   // mPanels[repoId]->PopulateOrExchange(S);
+
+   wxSizer *sizer = mScroller->GetSizer();
+   if (sizer)
+   {
+      sizer->SetSizeHints(mScroller);
+   }
+   mScroller->FitInside();
+   mScroller->Layout();
+   mScroller->GetParent()->Layout();
+}
+
+CardFetchedCallback ModelManagerPanel::GetCardFetchedCallback()
+{
+   CardFetchedCallback onCardFetched = [this](bool success, ModelCardHolder card)
+   {
+      this->CallAfter(
+         [this, success, card]()
+         {
+            if (success)
+            {
+               bool found = mPanels.find(card->GetRepoID()) != mPanels.end();
+               bool effectTypeMatches = card->effect_type() == mEffect->GetDeepEffectID();
+               if (!found && effectTypeMatches)
+                  this->AddCard(card);
+            }
+         }
+      );
+   };
+
+   return onCardFetched;
+}
+
+void ModelManagerPanel::FetchCards()
+{
+   DeepModelManager &manager = DeepModelManager::Get();
+
+   CardFetchedCallback onCardFetched = GetCardFetchedCallback();
+
+   CardFetchProgressCallback onCardFetchedProgress = [this](int64_t current, int64_t total)
+   {
+      this->CallAfter(
+         [this, current, total]()
+         {
+            if (mTools)
+               this->mTools->SetFetchProgress(current, total);
+         }
+      );
+   };
+
+   manager.FetchModelCards(onCardFetched, onCardFetchedProgress);
+   manager.FetchLocalCards(onCardFetched);
+}
+
+void ModelManagerPanel::SetSelectedCard(ModelCardHolder card)
+{
+   wxSize oldSize = GetSize();
+   wxSize oldScrollerSize = mScroller->GetSize();
+   int oldPos = mScroller->GetScrollPos(wxVERTICAL);
+   // set all other card panels to disabled
+   for (auto& pair : mPanels)
+   {
+      pair.second->SetModelStatus(ModelCardPanel::ModelStatus::Disabled);
+
+      if (card)
+      {
+         if (pair.first == card->GetRepoID())
+            pair.second->SetModelStatus(ModelCardPanel::ModelStatus::Enabled);
+      }
+   }
+
+   // configure the detailed panel
+   if (card)
+   {
+      mDetailedPanel->PopulateWithNewCard(card);
+      mDetailedPanel->SetModelStatus(ModelCardPanel::ModelStatus::Enabled);
+   }
+
+   SetSize(oldSize);
+   mScroller->SetSize(oldScrollerSize);
+   mScroller->Refresh();
+   mScroller->Scroll(0, oldPos);
+   // mScroller->FitInside();
+   // mScroller->Layout();
+   // mScroller->GetParent()->Layout();
+}
+
+// ManagerToolsPanel
+
+ManagerToolsPanel::ManagerToolsPanel(wxWindow *parent, ModelManagerPanel *panel)
+   : wxPanelWrapper((wxWindow *)parent, 
+                     wxID_ANY, 
+                     wxDefaultPosition, 
+                     getManagerToolsPanelSize())
+{
+   mManagerPanel = panel;
```

Comment:
I prefer to see member initialization before the constructor body where possible, not as assignments.  (In fact you have no choice when a member is const or is a reference.)  Or, as mentioned, initializing in-line in the class.  Again all this helps a reviewer check correctness of things.

Reply:
525fb9b9e497bcd38a812067cf7ce69350e9bcec